### ISSUE TITLE
Issues with Answer X of Y page type (model=topXQ.html)

### ISSUE THREAD

**Issue by ronm123** [thexerteproject/xerteonlinetoolkits#917]
Hi all
we're working through various page types whilst drafting an accessibility statement and I'm adding issues here labelled Accessibility (as well as any other relevant labels) and some are just general bugs/fixes needed that aren't necessarily accessibility specific but cropping up during testing. So the first is a couple of issues with the Answer X of Y page:

Accessibility checks reveal that labels are needed for the text entry form fields. I've added aria-label="answer '+(i+1)+'" and will commit that change but I'm conscious that 'answer' should be a language string. However as this would be an aria label I'm not sure if this should be added to the optional language strings in the wizard or picked up from a language pack? If the latter what existing file?

The page also currently shows null: null until an attempt has been made. I'm not sure what is the best way to resolve that?

**Comment by torinfo** [comment]
The only option to make it translatable now is to make it a optional language string. It might be possible to just add the string to the default parameteres, without adding a node in the xwd. I don't know if it gets picked up. I'll test that. If we are bale to do that, we can put it in the xwd, it is possible to translate it, but it will not appear as a language strin in the wizard. Like I said, I'll test that.

**Comment by torinfo** [comment]
No. You really have to add a language option. Otherwise it's not visisble for Xenith

**Comment by ronm123** [comment]
@torinfo thanks Tom. So I've added it to topXQ.xwd and I'm checking if the value is undefined or null in the model and making it "Answer" if so but as it's a new language variable it doesn't appear as a language text box in an existing Answer X of Y page only when you add a new Answer X of Y page. Remind me do we have a solution for that so that it appears as a language string option in existing pages too.

**Comment by ronm123** [comment]
I've tried adding mandatory="true" to the end e.g. <answerFieldLabel label="Answer Field Aria-Label" type="TextInput" wysiwyg="false" language="true" mandatory="true"/> and running rebuild but that doesn't seem to be working. But a quick search didn't reveal other language nodes with mandatory in them so perhaps that isn't enable for language nodes or am I missing something stupid?

**Comment by ronm123** [comment]
@torinfo I've committed this https://github.com/thexerteproject/xerteonlinetoolkits/commit/6553cc99c3d264ec993c2b2e6e3590efb292baad but not sure how to make the language option appear for existing pages.

Also will leave you to deal with the null null issue unless you can summarise what is needed to fix that?

**Comment by torinfo** [comment]
@ronm123 Will have a look at the language options and the null null issue.

**Comment by torinfo** [comment]
@ronm123 Fixed your commit about the language option in 7bd8c9fab2742601fa37c9c0c69c810d78154555

I don't know if it will always show for existing topXofY instances, it should. You can force it by selecting a different language, and then reselecting the language. I'll check why it doesn't always seem to work with the madatory is true. If you use mandatory="true" alwaus add a defaultValue too.

**Comment by ronm123** [comment]
@torinfo thanks Tom. I pulled down your commit and cleared cache but the language option still didn't appear on my existing page. However temporarily changing language to a different language and then back again did then reveal the new options in the existing page so at least there is a solution for those who want to change the default value on existing pages.

I'm not sure what your intention was with changing the code so that the default value is Answer {i}? Firstly I wouldn't imagine many people will change the default anyway or know where it appears but also I'm not sure we should provide control over the {i}. MY intention was that whatever an author added as the word for Answer each field would become Answer 1, Answer 2 and so on. That was working for me but with your changes they are all showing as Answer 1?

**Comment by torinfo** [comment]
@ronm123 It is an automatic reaction from me that wherever there are counters or constructs like {i} of {n} to use these replacement strings, as the translated strings in a different language sometimes need a different ordering of words.. I admit, that they are not needed in this case, and if you think it will lead to confusion I will remove it. Let me know. Also fixed the issue with all of them being Answer 1, but have not committed that yet, until I hear back from you.

**Comment by ronm123** [comment]
@torinfo although I would normally agree to put everything under author control I think in this case we would have to add a tooltip to explain what the {i} is representing and achieving and I'm not sure that's worth the risk of someone just replacing the whole string or leaving it empty. e.g. if someone removes the whole sting because they don't understand it or think it appears visible on screen the code should still work and assign an incremental label to each field. 

Actually as I typed that I think we need to add a tooltip anyway. So if you commit your change as it is with the {i} still in place (or if you think we should then remove the {i}) I'll have a go at adding a tooltip and possibly changing the label that appears in the wizard e.g. Answer Field Aria-Label so that it still displays ok with the help icon added e.g. make it shorter or longer so that it wraps.

**Comment by torinfo** [comment]
@ronm123 I have committed this change. It's 'Answer' at the moment. Be aware that existing LO's will still have Answer {i}, so the label at the moement is Answer {i} 1, Answer {i} 2, if you don't reset it.

**Comment by ronm123** [comment]
Thanks @torinfo I'll reset my test LO and obviously nobody else will have used this yet apart from us.

The bad news is that we have similar issues with all the other page types where we have text entry including dialogue, dictation, documentation, gap fill, model answer and modify exercise. None of those have text box labels and I guess some will be a simple fix where it's just a single text box but others like documentation may be more complex to fix. Before we start working on these individually is there another way to deal with this rather than working on each model and wizard separately?

**Comment by torinfo** [comment]
@ronm123 I am afraid not....

**Comment by ronm123** [comment]
@torinfo and I was afraid you were going to say that! ;-)

**Comment by ronm123** [comment]
Closing this for the Answer X of Y page but will open a new issue to check off the others as we do them.
